### PR TITLE
Allow specifying ORDER BY clause that uses a different ROW_TIMESTAMP …

### PR DESCRIPTION
**Pull request by twdsilva**
…column than the baset table while creating materialized views

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: adapter/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -64,6 +64,7 @@ SqlNode SqlCreateMaterializedView() :
      SqlIdentifier cubeName;
      boolean ifNotExists = false;
      SqlSelect query;
+     SqlNodeList orderList;
```
Comment (rbhanot) [inline adapter/src/main/codegen/includes/parserImpls.ftl:67]: Have lost the context I have over the grammar, so trying to understand the need for a SqlNodeList ? 
Is this grammar representation for all Materialized Views of a dataset together and not per view ? (and hence the list)

Comment (twdsilva) [reply adapter/src/main/codegen/includes/parserImpls.ftl:67]: `SqlSelect` doesn't allow you to specify an `ORDER BY`. It only allows a basical SELECT with `GROUP BY`.

**Review by rbhanot — APPROVED** [review]

File: adapter/src/main/java/com/twilio/kudu/sql/KuduQuery.java [hunk 2]
```diff
@@ -32,6 +32,7 @@
 import org.apache.calcite.rel.metadata.JaninoRelMetadataProvider;
 import org.apache.calcite.rel.metadata.RelMetadataQuery;
 import org.apache.calcite.rel.rules.CoreRules;
+import org.apache.calcite.rel.rules.ProjectRemoveRule;
```
Comment (minakshikorad) [inline adapter/src/main/java/com/twilio/kudu/sql/KuduQuery.java:35]: nit: why do we need this import?

File: adapter/src/main/java/com/twilio/kudu/sql/parser/SqlCreateMaterializedView.java [hunk 3]
```diff
@@ -35,14 +36,17 @@ public class SqlCreateMaterializedView extends SqlCall {
   public final SqlIdentifier cubeName;
   public final boolean ifNotExists;
   public final SqlSelect query;
+  public final SqlNodeList orderByList;
 
   /** Creates a SqlCreateView. */
-  SqlCreateMaterializedView(SqlParserPos pos, SqlIdentifier cubeName, boolean ifNotExists, SqlSelect query) {
+  SqlCreateMaterializedView(SqlParserPos pos, SqlIdentifier cubeName, boolean ifNotExists, SqlSelect query,
+      SqlNodeList orderByList) {
```
Comment (minakshikorad) [inline adapter/src/main/java/com/twilio/kudu/sql/parser/SqlCreateMaterializedView.java:43]: will the kudu range partitions also be setup based on this column?
do we need changes to kk while setting up the partitioning?

File: adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java [hunk 4]
```diff
@@ -118,6 +111,29 @@ public void clearCachedTableMap() {
...
+    try {
+      KuduTable kuduTable = this.client.openTable(tableName).join();
+      for (ColumnSchema columnSchema : kuduTable.getSchema().getColumns()) {
+        String comment = columnSchema.getComment();
+        if (!comment.isEmpty()) {
+          JSONObject jsonObject = new JSONObject(comment);
+          if (jsonObject.has("isTimeStampColumn") && jsonObject.getBoolean("isTimeStampColumn")) {
+            timeStampColumnName = columnSchema.getName();
+          }
+          if (jsonObject.has("isDescendingSortOrder") && jsonObject.getBoolean("isDescendingSortOrder")) {
```
Comment (minakshikorad) [inline adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java:126]: would we need this check only if it is a `isTimeStampColumn` is true?

File: adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java [hunk 5]
```diff
@@ -258,14 +261,34 @@ protected Map<String, Table> getTableMap() {
           }
           queryBuilder.append(Util.toString(groupByExprs, "", ", ", ""));
 
+          // TODO figure out why adding the ORDER BY leads to the view not being chosen
```
Comment (minakshikorad) [inline adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java:264]: nit: remove commented lines

File: adapter/src/main/java/com/twilio/kudu/sql/KuduQuery.java (hunk 2 above)
Comment (twdsilva) [reply adapter/src/main/java/com/twilio/kudu/sql/KuduQuery.java]: reverted

File: adapter/src/main/java/org/apache/calcite/prepare/KuduPrepareImpl.java [hunk 6]
```diff
@@ -366,30 +462,23 @@ else if (kuduTable.getSchema().hasColumn(sqlnode.toString())) {
...
-            } else {
-              lowerBound.addTimestamp(rowTimestampColumn, new Timestamp(Long.MIN_VALUE));
-              upperBound.addTimestamp(rowTimestampColumn, new Timestamp(Long.MIN_VALUE + 1));
-            }
+        // create a single dummy range partition using the rollup timestamp column
+        PartialRow lowerBound = cubeSchema.newPartialRow();
+        PartialRow upperBound = cubeSchema.newPartialRow();
+        if (kuduSchema.createDummyPartition) {
+          // Set range partition to EPOCH.MAX - minvalue for DESC case.
+          if (descendingCols.contains(rollupTimestampColumn)) {
```
Comment (twdsilva) [inline adapter/src/main/java/org/apache/calcite/prepare/KuduPrepareImpl.java:470]: The range partitioning of the cube is determined by the timestamp column on which the roll up is being done. As long as the base table and cube tables have stores timestamps that are close then range partitioning on different columns should be ok. If there is a large difference in time between the timestamp column of the fact and cube there could be data inconsistency issues while dropping range partitions based on time.

File: adapter/src/main/java/com/twilio/kudu/sql/parser/SqlCreateMaterializedView.java (hunk 3 above)
Comment (twdsilva) [reply adapter/src/main/java/com/twilio/kudu/sql/parser/SqlCreateMaterializedView.java:43]: I pointed out in a separate comment where how the range partitioning of cubes are setup.

File: adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java (hunk 4 above)
Comment (twdsilva) [reply adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java:126]: There can be more than one descending column, but there can be only one timestamp column for eg
```
ORDER BY STRING_COL, FLOOR(TIMESTAMP_COLUMN2 TO hour) DESC, INT32_COLUMN2 DESC
```

File: adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java (hunk 5 above)
Comment (twdsilva) [reply adapter/src/main/java/com/twilio/kudu/sql/schema/KuduSchema.java:264]: Ideally while creating the SQL statement that specified the materialized view should also have the ORDER BY clause, but for some reason when I add this the view isn't chosen for queries. Currently, the view sql only has the GROUP BY clause and the KuduSortRule checks the sort order of the query vs the view and is matched. 
I left this code in with the TODO to look into this.

**Review by minakshikorad — APPROVED** [review]

**Pull request merged**